R_notebooks/05_Combining_datasets.Rmd

```diff
@@ -0,0 +1,214 @@
+the columns from `data_left`. We bring in as many rows from `data_right` as we
+can. The 'left' in `left_join` indicates that we're keeping everything from the
+'left' dataframe (i.e. the first one) and joining the other dataframe onto the
+'left' one.
+
+It's helpful to use small, toy dataframes to explore how joins work (or remind
+yourself after a period of time away).
+
+```{r}
+df1 <- data.frame(colA = c(1, 2, 3, 4),
+                  colB = c(2021, 2022, 2023, 2024),
+                  colC = c('a', 'b', 'c', 'd'))
+
+df2 <- data.frame(colA = c(1, 3),
+                  colD = c('foo', 'bar'),
+                  colE = c('dog', 'cat'))
+```
+
+```{r}
+df1
+```
+
+```{r}
+df2
+```
+
+Now let's join `df2` onto `df1` by the column `colA`:
+
+```{r}
+df1 |>
+  dplyr::left_join(df2, by = "colA")
+```
+
+Notice:
+
+- All the data from the left-hand dataframe, `df1`, is kept (columns `colA` -- `colC`).
+
+- The data from the right-hand dataframe `df2` has been brought over
+  for the rows where the 'by' column `colA` has matching values in the
+  left-hand dataframe `df1`.
+  
+- The rows where there aren't matching values in the 'by' column `colA` have
+  missing values for the other right-hand dataframe columns (`colD` and `colE`
+  in this case).
+
+
+### Exercise: left join practice
+
+Suppose now the right hand dataframe is as follows:
+
+```{r}
+df2 <- data.frame(colA = c(1, 3, 2, 4, 5),
+                  colD = c('foo', 'bar', 'baz', 'qux', 'foo'),
+                  colE = c('dog', 'cat', 'mouse', 'rabbit', 'horse'))
+```
+
+Can you guess what the output of `dplyr::left_join(df1, df2, by = "colA")` will
+be? Check your guess with code below.
+
+```{r}
+df1 |>
+  dplyr::left_join(df2, by = "colA")
+```
+
+Now what if you swap the order of `df1` and `df2`? Guess the output of
+`dplyr::left_join(df2, df1, by = "colA")` and check your answer below.
+
+```{r}
+df2 |>
+  dplyr::left_join(df1, by = "colA")
+```
+
+
+### Optional exercise: repeated keys
+
+The values in the 'by' column(s) are sometimes called _keys_ for the join. The
+rules we described in the above example are not the whole story and can be
+complicated by the presence of repeated keys.
+
+What do you think happens if there are repeated keys? For example, try to
+guess the output of the following code:
+
+```{r}
+df1 <- data.frame(colA = c(1, 2, 3, 3),  # repeated key 3
+                  colB = c(2021, 2022, 2023, 2024),
+                  colC = c('a', 'b', 'c', 'd'))
+
+df2 <- data.frame(colA = c(1, 3),
+                  colD = c('foo', 'bar'),
+                  colE = c('dog', 'cat'))
+
+df1 |>
+  dplyr::left_join(df2, by = "colA")
+```
+
+Now what if we swap the roles of `df1` and `df2`?
+
+```{r}
+df2 |>
+  dplyr::left_join(df1, by = "colA")
+```
+
+#### Warning about doubles
+
+It's generally not a good idea to join on a 'by' column (or columns)
+that contain doubles. This is because matching in the join will be done
```

Comment:
Consider rewording "that contain doubles" to "that are of type double" - for ease of understanding